File: betterrandom/src/test/java/io/github/pr0methean/betterrandom/prng/adapter/SplittableRandomAdapterTest.java
package io.github.pr0methean.betterrandom.prng.adapter;

import static io.github.pr0methean.betterrandom.seed.DefaultSeedGenerator.DEFAULT_SEED_GENERATOR;
import static io.github.pr0methean.betterrandom.seed.SecureRandomSeedGenerator.DEFAULT_INSTANCE;
import static org.testng.Assert.assertEquals;
import static org.testng.Assert.assertNotEquals;

import com.google.common.testing.SerializableTester;
import io.github.pr0methean.betterrandom.FlakyRetryAnalyzer;
import io.github.pr0methean.betterrandom.prng.BaseRandom;
import io.github.pr0methean.betterrandom.prng.RandomTestUtils;
import io.github.pr0methean.betterrandom.prng.RandomTestUtils.EntropyCheckMode;
import io.github.pr0methean.betterrandom.seed.FakeSeedGenerator;
import io.github.pr0methean.betterrandom.seed.RandomSeeder;
import io.github.pr0methean.betterrandom.seed.SeedException;
import io.github.pr0methean.betterrandom.seed.SeedGenerator;
import io.github.pr0methean.betterrandom.seed.SemiFakeSeedGenerator;
import io.github.pr0methean.betterrandom.util.BinaryUtils;
import java.util.concurrent.ThreadLocalRandom;
import org.testng.annotations.AfterMethod;
import org.testng.annotations.BeforeMethod;
import org.testng.annotations.Test;

public class SplittableRandomAdapterTest
    extends SingleThreadSplittableRandomAdapterTest {

  private RandomSeeder thread;

  @BeforeMethod public void setUp() {
    thread = new RandomSeeder(getTestSeedGenerator());
  }

  @AfterMethod public void tearDown() {
    thread.stopIfEmpty();
    thread = null;
  }

  @Override protected EntropyCheckMode getEntropyCheckMode() {
    return EntropyCheckMode.LOWER_BOUND;
  }

  @Override protected SplittableRandomAdapter createRng() throws SeedException {
    return new SplittableRandomAdapter(getTestSeedGenerator(), thread);
  }

  @Override protected BaseRandom createRng(byte[] seed) throws SeedException {
    SplittableRandomAdapter out = createRng();
    out.setSeed(seed);
    return out;
  }

  // FIXME: Why does this need more time than other PRNGs?!
  @Test(timeOut = 120_000) @Override public void testDistribution() throws SeedException {
    super.testDistribution();
  }

  @Override public void testInitialEntropy() {
    // This test needs a separate instance from all other tests, but createRng() doesn't provide one
    SplittableRandomAdapter random = createRng();
    assertEquals(random.getEntropyBits(), Long.SIZE, "Wrong initial entropy");
  }

  // FIXME: Why does this need more time than other PRNGs?!
  @Test(timeOut = 120_000) @Override public void testIntegerSummaryStats() throws SeedException {
    super.testIntegerSummaryStats();
  }

  @Override @Test public void testSerializable() throws SeedException {
    try {
      RandomSeeder thread = new RandomSeeder(DEFAULT_SEED_GENERATOR);
      final BaseSplittableRandomAdapter adapter =
          new SplittableRandomAdapter(DEFAULT_INSTANCE, thread);
      final BaseSplittableRandomAdapter clone = SerializableTester.reserialize(adapter);
      assertEquals(adapter, clone, "Unequal after serialization round-trip");
    } finally {
      thread.shutDown();
    }
  }

  @Override protected Class<? extends BaseRandom> getClassUnderTest() {
    return SplittableRandomAdapter.class;
  }

  @Override @Test(enabled = false) public void testRepeatability() {
    // No-op.
  }

  @Override @Test(enabled = false) public void testRepeatabilityNextGaussian() {
    // No-op.
  }

  @Override @Test(retryAnalyzer = FlakyRetryAnalyzer.class)
  public void testReseeding() {
    SeedGenerator generator = new SemiFakeSeedGenerator(ThreadLocalRandom.current(), "testReseeding");
    RandomSeeder seeder = new RandomSeeder(generator);
    try {
      SplittableRandomAdapter random = new SplittableRandomAdapter(generator, seeder);
      RandomTestUtils.checkReseeding(generator, random, false);
    } finally {
      seeder.shutDown();
    }
  }

  /**
   * Test for crashes only, since setSeed is a no-op.
   */
  @Override @Test public void testSetSeedAfterNextLong() throws SeedException {
    final BaseRandom prng = createRng();
    prng.nextLong();
    prng.setSeed(getTestSeedGenerator().generateSeed(8));
    prng.setSeed(BinaryUtils.convertBytesToLong(getTestSeedGenerator().generateSeed(8)));
    prng.nextLong();
  }

  /**
   * Test for crashes only, since setSeed is a no-op.
   */
  @Override @Test public void testSetSeedAfterNextInt() throws SeedException {
    final BaseRandom prng = createRng();
    prng.nextInt();
    prng.setSeed(getTestSeedGenerator().generateSeed(8));
    prng.setSeed(BinaryUtils.convertBytesToLong(getTestSeedGenerator().generateSeed(8)));
    prng.nextInt();
  }

  /**
   * Assertion-free since reseeding may cause divergent output.
   */
  @Override @Test(timeOut = 10000) public void testSetSeedLong() {
    createRng().setSeed(0x0123456789ABCDEFL);
  }

  /**
   * setRandomSeeder doesn't work on this class and shouldn't pretend to.
   */
  @Override @Test(expectedExceptions = UnsupportedOperationException.class)
  public void testRandomSeederThreadIntegration() {
    RandomSeeder thread = new RandomSeeder(DEFAULT_INSTANCE);
    try {
      createRng().setRandomSeeder(thread);
    } finally {
      thread.stopIfEmpty();
    }
  }

  @Test public void testSetSeedGeneratorNoOp() {
    createRng().setRandomSeeder(thread);
  }

  @Override @Test(enabled = false) public void testSeedTooShort() {
    // No-op.
  }

  @Override @Test(enabled = false) public void testSeedTooLong() {
    // No-op.
  }

  @Override @Test public void testDump() throws SeedException {
    RandomSeeder thread = new RandomSeeder(DEFAULT_INSTANCE);
    try {
      SplittableRandomAdapter firstInstance
          = new SplittableRandomAdapter(getTestSeedGenerator(), thread);
      RandomSeeder otherThread =
          new RandomSeeder(new FakeSeedGenerator("Different reseeder"));
      try {
        SplittableRandomAdapter secondInstance =
            new SplittableRandomAdapter(getTestSeedGenerator(), otherThread);
        assertNotEquals(secondInstance.dump(), firstInstance.dump());
      } finally {
        otherThread.shutDown();
      }
    } finally {
      thread.shutDown();
    }
  }

  /**
   * Assertion-free because thread-local.
   */
  @Override @Test public void testThreadSafety() {
    testThreadSafetyVsCrashesOnly(45, functionsForThreadSafetyTest);
  }
}

File: betterrandom/src/test/java/io/github/pr0methean/betterrandom/prng/adapter/SplittableRandomAdapterTest.kt
package io.github.pr0methean.betterrandom.prng.adapter

import org.testng.Assert.assertEquals

import io.github.pr0methean.betterrandom.TestingDeficiency
import io.github.pr0methean.betterrandom.prng.BaseRandom
import io.github.pr0methean.betterrandom.prng.RandomTestUtils
import io.github.pr0methean.betterrandom.seed.SeedException
import org.testng.annotations.Test

class SplittableRandomAdapterTest : SingleThreadSplittableRandomAdapterTest() {

    protected override val classUnderTest: Class<out BaseRandom>
        get() = SplittableRandomAdapter::class.java

    /** SplittableRandomAdapter isn't repeatable until its seed has been specified.  */
    @Throws(SeedException::class)
    override fun testRepeatability() {
        val rng = createRng()
        rng.setSeed(BaseRandomTest.TEST_SEED)
        // Create second RNG using same seed.
        val duplicateRNG = createRng()
        duplicateRNG.setSeed(BaseRandomTest.TEST_SEED)
        RandomTestUtils.assertEquivalent(rng, duplicateRNG, 1000, "Generated sequences do not match")
    }

    /** SplittableRandomAdapter isn't repeatable until its seed has been specified.  */
    @TestingDeficiency // Failing
    @Test(enabled = false)
    @Throws(SeedException::class)
    override fun testRepeatabilityNextGaussian() {
        val rng = createRng()
        val seed = testSeedGenerator.generateSeed(getNewSeedLength(rng))
        rng.nextGaussian()
        rng.seed = seed
        // Create second RNG using same seed.
        val duplicateRNG = createRng()
        duplicateRNG.seed = seed
        assertEquals(rng.nextGaussian(), duplicateRNG.nextGaussian())
    }

    @Throws(SeedException::class)
    override fun createRng(): SplittableRandomAdapter {
        return SplittableRandomAdapter(testSeedGenerator)
    }

    /** Seeding of this PRNG is thread-local, so setSeederThread makes no sense.  */
    @Test(expectedExceptions = arrayOf(UnsupportedOperationException::class))
    @Throws(Exception::class)
    override fun testRandomSeederThreadIntegration() {
        createRng().seedGenerator = testSeedGenerator
    }

    /** Assertion-free because thread-local.  */
    @Test
    override fun testThreadSafety() {
        testThreadSafetyVsCrashesOnly(30, BaseRandomTest.FUNCTIONS_FOR_THREAD_SAFETY_TEST)
    }
}
